### ISSUE TITLE
Running on other datasets

### ISSUE THREAD

**Issue by iqkhilji** [microsoft/CryptoNets#7]
Hello,
So ,we know that the  CryptoNets project is trained with MNIST dataset and works on currently 3 different datasets namely **MNIST**, **CIFAR-10** and **CALTECH-101**. 
My question is , can it run on any other datasets?. I am having my own dataset and want to train the CrytoNets model with it and run the model. So, is there any way to do it ? Thank you

**Comment by ranigb** [comment]
It is possible to run on other datasets. We are in the process of documenting the process in the Wiki of this project under the  [bring your own model](https://github.com/microsoft/CryptoNets/wiki/Bring-Your-Own-Model). The documentation is not complete yet but this is WIP. We will be happy to take comments and suggestions.

**Comment by iqkhilji** [comment]
Thank you very much. Looking forward to the last steps of building the application (Step 4,5 and 6).

**Comment by iqkhilji** [comment]
Hello again,
So i was able to train and create a model for deep neural net . But I am having some confusions. Firstly in the part **'Converting weights to CryptoNets format'** , is the code for **all the layers** all together or jst for a **single layer.**  as it generates 2 files (bias.csv and weight.csv)  when run but later it is told that to sum all the weights and all the biases (weight1.csv ,weight2.csv...... bias1.csv,bias2.csv etc). How would i know that how many layers it is using and how would i produce the other weights and bias files by modifying the given code? 

Secondly , when i try to build the application in Microsoft Visual studio 2019 . The following errors show up . ' using HE Wrapper is underlined red and says "The type of name 'HE Wrapper could not be found(are you missing a using directive or an assembly reference?)". This is the error for 'using NeuranNetworks' and many other functions like 'RawFactory' , 'WeightsReader'  etc. 
![application comlpaint](https://user-images.githubusercontent.com/39790947/65462309-5f214080-de77-11e9-88d9-af9d5a2e8ccf.PNG)
 
Another error is in the line of code 'Weights = ((double[])wr.Weights[1]) ,
                                                          Bias = ((double[])wr.Biases[1]),' 
 where it says  "cannot apply indexing with [] to an expression of type object."
![Complaint 2](https://user-images.githubusercontent.com/39790947/65464042-ca204680-de7a-11e9-9bea-8a5bc78a0934.PNG)

Third  , in the line of code ' Source = DenseLayer4'  it says "The name 'DenseLayer4' does not exist in the current context"
![cmplaint 4](https://user-images.githubusercontent.com/39790947/65464370-69453e00-de7b-11e9-9f8f-579c9af5fdd5.PNG)

Lastly, in the line of code  ' Utils.ProcessInEnv(env =>"  ,  the error is " 'Utils' does not contain a definition for 'ProcessInEnv' "
![complaint 3](https://user-images.githubusercontent.com/39790947/65464802-92b29980-de7c-11e9-9e3f-cecb3ccf0c66.PNG)

Can you please tell me what could be causing these problems ? My confusion at first are the issue with blocks and sum of the weights  and when i try to create the application using jst the two files of 'weights.csv' and  'bias.csv'  these errors appear. There are 16 errors in total and a warning.  I have built my own training model but am finding it difficult to build the application.  Please help me out.
Thank you

**Comment by ranigb** [comment]
You have multiple questions here, I'll try to address them one by one:
1. For each layer with trainable weights (a dense layer or a convolution layer) you generate a bias file and a weights file. Once you have done that for all the relevant layers, you combine all the weights into a one file and all the biases into a second file. This is described at the end of  [this section](https://github.com/microsoft/CryptoNets/wiki/Bring-Your-Own-Model#converting-weights-to-cryptonets-format) where it suggests using the commands:
```
gawk ‘{print $0;}’ weights1.csv weights2.csv weights3.csv > all_weights.csv 
```
and
```
gawk ‘{print $0;}’ bias1.csv bias2.csv bias3.csv > all_bias.csv 
```

2. You received an error message about HEWrapper missing. Did you add a reference to the HEWrapper package in the references section of your project?

3. Regarding the errors in lines such as `Weights = ((double[])wr.Weights[1])`: can you show the line where you generate the wr object and load the weights? It will also help to see what type visual studio thinks the wr object is and the type of wr,Weights

4. For the error DenseLayer4 does not exists, can you show where DenseLayer4 is defined?

5. The error on Utils.ProcessInEnv: this is likely because of the missing HEWrapper package.


I this response helps.

**Comment by iqkhilji** [comment]
By the layers you mean the layers with trainable weights [128,5,5] , [83,3,3] and [130,5,5] ? if so , then is the input_shape of all the layers will be {83,14,14} which is for ConvLayer3 in section 'Create your own model?

Will adding the references of the dll files namely 'NeuralNetworks.dll' (in the CryptoNets-master\NeuralNetworks\obj\x64\Release directory) , 'HE Wrapper.dll' (in the \CryptoNets-master\HE Wrapper\obj\x64\Release)   and  'SEALNet.dll'  in the (SEAL\dotnet\src\obj\Release\netstandard2.0)  for libraries of NeuralNetworks, HE Wrapper and SEAL help in the cases ?

And as for the 'DenseLayer4' i copy pasted the code that was given in that section. If i were to define Denselayer4 , can you please mention what would be the necessary parameters of this layer?

Thank You

**Comment by ranigb** [comment]
Again, lets work the answers to your questions one by one:
1. The input to each layer has different size. The input shape depends on the network structure that you use. In Keras, you can use the command model.summary() to find the shape of each layer

2. Indeed, you have to add 'NeuralNetworks.dll',  'HE Wrapper.dll' and 'SEALNet.dll' as references

3. As far as I can see, the code in the wiki does not contain DenseLayer4. However you seem to have implemented it in your code - this is perfectly fine. If you didn't want to add another layer to the network you should point to one of the existing layers. If adding a layer was done by design then note that you are using the same weights and biases for DenseLayer4 as you do for ConvLayer3 which is not very likely to work as they have different shapes. To get the parameters you have to train a model that has this layer and save its parameters so you can use them in DenseLayer4

4. "test.txt" is a file that you should supply with test example(s). If you are using the example code then test.txt should have one line per image where each line contains 1 + 3*32*32 tab separated columns in which the first column is the label and the other column are the RGB values of a 32*32 image. You can follow the instructions to obtain the CIFAR dataset to see how such file should look like.

**Comment by iqkhilji** [comment]
Thanks a lot for your response. 
I have managed to get the total number of layers using the commands that you provided and i have found 14 layers among which 5 have parameters and others showing 0 parameters . So i think i should compute the weights and biases of those layers only.

But meanwhile , I am facing other issues. I have run the code  given in the wiki for weights and biases which generates 'weights.csv' and 'bias.csv' file for a single layer i think . I know i should have computed the sum of the weights and biases for all the layers and  passes on 'all_bias.csv' and 'all_weights .csv'   in the line of code  **WeightsReader wr = new WeightsReader("weights.csv", "bias.csv");** instead of jst weights and bias but wanted to try this instead for single layer at first. 
I have also taken the 'cifar-test.csv' file from the given pre-trained CryptoNets model provided here and converted it to 'test.txt' file which I am using in the code  **string fileName = "test.txt";**  since it may already has what you just said about the 'text.txt' file earlier.  Below is the picture 
![weights for 3d](https://user-images.githubusercontent.com/39790947/65618483-281a6e80-dfe0-11e9-9586-aa4cfc155ba0.PNG)

However, I am now facing some issues yet again. When I start to run my program, there is an exception which says **System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection. Parameter name: index'** along with errors 
**arraylist.cs not found**  and **Symbol file not loaded** . Even if i make a separate file for 'arraylist.cs' with the necessary code , there is still this exception. Below are the pictures.
![arraylistcs](https://user-images.githubusercontent.com/39790947/65619065-4c2a7f80-dfe1-11e9-8267-45e21d1b75c6.PNG)
![symbolfile](https://user-images.githubusercontent.com/39790947/65619074-4fbe0680-dfe1-11e9-845f-27300ce6486a.PNG)

Thanks a ton for your help till now. I hope you'll keep helping me till I finally successfully run my own model . Thanks again.

**Comment by ranigb** [comment]
It looks as if these problems are rooted in some format issues in weights.csv and bias.csv. Probably the easiest thing to do would be to start with one of the existing applications. Actually, the code in wiki describes how to build the model for CIFAR that is in the source code. Have a look at this code and in the weights and biases files provided as CifarWeight.csv and CifarBias.csv.

One thing that is important to clarify is that after you generate the weights for each layer you do NOT sum these weights: instead you should combine them into a single file such that each line of this file contains the weights from a single layer. You do the same thing for the biases.

I think that walking through one of the existing examples could help and the CIFAR one is a good starting point since the wiki describes it in details.

**Comment by iqkhilji** [comment]
Thank You for your response again

Let me clarify by telling you that these I have tried to build the model using Cifar datasets in the first place all along . I was planning to run my own datasets afterwards once I had build the model and run it with Cifar datasets sucessfully. But these issues kept arising whatsoever. 

Hence, as i have mentioned earlier I took the 'cifar_test.csv' file from the CrytoNets existing pre-trained model that you provided and converted into 'test.txt' to use it in the code from the wiki and incase of 'weights.csv' and 'bias.csv' , i've used those from the 'cifar_model.h5' model that i've built with Cifar dataset using the code from the wiki. I think that these files (weights.csv and bias.csv) are of a single layer tho, but wanted to try it in the wiki code before combining all the files of weights and biases from each layer and using those files (all_weights.csv ,all_bias.csv) and using it. 

I think that the issue maybe is in fact using the files 'weights.csv' and 'bias.csv' which are maybe for a single layer that generated from running the code from the wiki and not obtaining the weights and biases for other layers and using them to form the file 'all_weights.csv' and  'all_bias,csv' and using them in the wiki code instead. 
Isn't it?. Thanks again till now .

**Comment by iqkhilji** [comment]
Hello,

I had a confusion regarding the 'Selecting the encryption parameters' section.  Is the code given in this section for 'BFV scheme' only or we can use it for 'CKKS scheme' as well?  If it is for 'CKKS scheme' too, will changing the line  **var Factory = new RawFactory((ulong)batchSize);  // we start without using encryption**   to     **var Factory = new EncryptedSealCkksFactory(new ulong[] { 957181001729, 957181034497 }, 16384);**   help in this case ? 

Thank you again

**Comment by iqkhilji** [comment]
Greeting Sir,

You'll be happy to know that i have successfully run my first image classification CNN model using my very own research dataset converted to cifar 10 format  which has only **2 classes to classify.**

However , i am now currently facing issues regarding obtaining the **weights and biases of each trainable layer** in my model and saving them in separate documents to combine them into a single file ('all weights' and 'all biases'). I have obtained how all the layers looks like by using 'model.summary()' like you said earlier. This is shown below


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_4 (Conv2D)            (None, 32, 32, 128)       3584      
_________________________________________________________________
average_pooling2d_4 (Average (None, 16, 16, 128)       0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 14, 14, 83)        95699     
_________________________________________________________________
dropout_3 (Dropout)          (None, 14, 14, 83)        0         
_________________________________________________________________
activation_4 (Activation)    (None, 14, 14, 83)        0         
_________________________________________________________________
average_pooling2d_5 (Average (None, 7, 7, 83)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 7, 7, 130)         269880    
_________________________________________________________________
activation_5 (Activation)    (None, 7, 7, 130)         0         
_________________________________________________________________
average_pooling2d_6 (Average (None, 3, 3, 130)         0         
_________________________________________________________________
dropout_4 (Dropout)          (None, 3, 3, 130)         0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 1170)              0         
_________________________________________________________________
dense_3 (Dense)              (None, 512)               599552    
_________________________________________________________________
dense_4 (Dense)              (None, 10)                5130      
_________________________________________________________________
activation_6 (Activation)    (None, 10)                0         
=================================================================

In the wiki, the code for generating 'weights.csv' and 'bias.csv'  for layer [14,14,83] is only given. But I am having trouble doing it for the other layers as well since when I input the tensor shape of layer  into 'input_shape=    '  keeping the kernel range same . It shows some error whatever i do.

Your help is again appreciated Sir,

Thank you

**Comment by engrmusawarali** [comment]
using System;
using NeuralNetworks;
using HEWrapper;


namespace CifarCryptoNet
{
    public class CifarCryptoNets
    {
        static void Main(string[] args)
        {
            string fileName = "cifar-test.tsv";
            int batchSize = 1;                        // number of examples to use per batch
            int numberOfRecords = 10;      // number of records in the test set
            var Factory = new RawFactory((ulong)batchSize);  // we start without using encryption
                                                             // this will make things faster
                                                             // and easier to debug

            WeightsReader wr = new WeightsReader("CifarWeight.csv", "CifarBias.csv");
            // load parameters

            var ReaderLayer = new BatchReader  // reads data from file
            {
                FileName = fileName,
                SparseFormat = false,
                MaxSlots = batchSize,
                NormalizationFactor = 1.0 / 256.0,
                Scale = 8.0
            };

            var EncryptedLayer = new EncryptLayer() { Source = ReaderLayer, Factory = Factory };

            var StartTimingLayer = new TimingLayer() { Source = EncryptedLayer, StartCounters = new string[] { "Batch-Time" } };

            // next we define the structure of the network


            var ConvLayer1 = new PoolLayer()
            {
                Source = StartTimingLayer,
                InputShape = new int[] { 3, 32, 32 },
                KernelShape = new int[] { 3, 8, 8 },
                Upperpadding = new int[] { 0, 1, 1 },
                Lowerpadding = new int[] { 0, 1, 1 },
                Stride = new int[] { 1000, 2, 2 },
                MapCount = new int[] { 83, 1, 1 },
                WeightsScale = 128000.0,
                Weights = (double[])wr.Weights[0],
                Bias = (double[])wr.Biases[0]
            };


            var ActivationLayer2 = new SquareActivation()
            {
                Source = ConvLayer1
            };


            var ConvLayer3 = new PoolLayer
            {
                Source = ActivationLayer2,
                InputShape = new int[] { 83, 14, 14 },
                KernelShape = new int[] { 8, 10, 10 },
              //  Upperpadding = new int[] { 0, 4, 4 },
                //Lowerpadding = new int[] { 0, 4, 4 },
                Stride = new int[] { 83, 2, 2 },
                MapCount = new int[] { 112, 1, 1 },
                WeightsScale = 1280000.0,
                Weights = ((double[])wr.Weights[1]),
                Bias = ((double[])wr.Biases[1])
            };


            var ActivationLayer4 = new SquareActivation()
            {
                Source = ConvLayer3
            };

            var DenseLayer5 = new PoolLayer()
            {
                InputShape = new int[] { 16384 },
                KernelShape = new int[] { 16384 },
                Source = ActivationLayer4,
                Weights = (double[])wr.Weights[2],
                Bias = (double[])wr.Biases[2],
                MapCount = new int[] { 10 },
                WeightsScale = 256000.0
            };


            var StopTimingLayer = new TimingLayer() { Source = DenseLayer5, StopCounters = new string[] { "Batch-Time" } };
            var network = StopTimingLayer;
            OperationsCount.Reset();
            Console.WriteLine("Preparing");
            network.PrepareNetwork();
            OperationsCount.Print();
            OperationsCount.Reset();
            for (var p = (INetwork)network; p != null; p = p.Source)
                if (p is BaseLayer b) b.Verbose = true;

            int count = 0;
            int errs = 0;
            while (count < numberOfRecords)
            {
                using (var m = network.GetNext())
                    Utils.ProcessInEnv(env =>
                    {
                        var decrypted = m.Decrypt(env);
                        for (int i = 0; i < decrypted.RowCount; i++)
                        {
                            int pred = 0;
                            for (int j = 1; j < decrypted.ColumnCount; j++)
                            {
                                if (decrypted[i, j] > decrypted[i, pred]) pred = j;
                            }
                            if (pred != ReaderLayer.Labels[i]) errs++;
                            count++;
                            if (count % 100 == 0)
                                Console.WriteLine("errs {0}/{1} accuracy {2:0.000}% prediction {3} label {4}", errs, count, 100 - (100.0 * errs / (count)), pred, ReaderLayer.Labels[i]);
                        }
                        Console.WriteLine("Batch size {0} {1}", batchSize, TimingLayer.GetStats());

                    }, Factory);
            }
            Console.WriteLine("errs {0}/{1} accuracy {2:0.000}%", errs, count, 100 - (100.0 * errs / (count)));
            network.DisposeNetwork();
        }
      }
}

**Comment by engrmusawarali** [comment]
This code is not working!!! I have tried my best and I am taking the weights provided by you for Lola Cifar CryptoNets. I would be thankful for your kind help.

**Comment by engrmusawarali** [comment]
Dear Ran,

I am able to run the code of Cifar10, but I am not getting the accuracy now. 

I am using the weights which are provided by you.
